xpti/include/xpti/xpti_trace_writer.hpp

```diff
@@ -0,0 +1,280 @@
+                   std::initializer_list<const char *> Categories,
+                   uint64_t Instance, trace_event_data_t *Event, size_t PID = 0,
+                   size_t TID = 0, size_t TimeStamp = 0) {
+    saveTracePoint(Type::End, EventName, std::move(Categories), Instance,
+                   /*Value=*/0, /*Parent=*/nullptr, Event, PID, TID, TimeStamp);
+  }
+  void addCounterEvent(const char *EventName,
+                       std::initializer_list<const char *> Categories,
+                       uint64_t Instance, uint64_t Value,
+                       trace_event_data_t *Event, size_t PID = 0,
+                       size_t TID = 0, size_t TimeStamp = 0) {
+    saveTracePoint(Type::Counter, EventName, std::move(Categories), Instance,
+                   Value, /*Parent=*/nullptr, Event, PID, TID, TimeStamp);
+  }
+  void addInstantEvent(const char *EventName,
+                       std::initializer_list<const char *> Categories,
+                       uint64_t Instance, trace_event_data_t *Event,
+                       size_t PID = 0, size_t TID = 0, size_t TimeStamp = 0) {
+    saveTracePoint(Type::Instant, EventName, std::move(Categories), Instance,
+                   /*Value=*/0, /*Parent=*/nullptr, Event, PID, TID, TimeStamp);
+  }
+
+  void connectEvents(trace_event_data_t *Parent, trace_event_data_t *Event) {
+    saveTracePoint(
+        Type::Connect, /*EventName=*/nullptr, /*Catecories=*/{}, /*Instance=*/0,
+        /*Value=*/0, Parent, Event, /*PID=*/0, /*TID=*/0, /*TimeStamp=*/0);
+  }
+
+  virtual ~TraceWriter() = default;
+
+protected:
+  enum class Type {
+    Begin,
+    End,
+    Counter,
+    Instant,
+    Connect,
+  };
+
+  struct TracePoint {
+    Type TracePointType;
+    const char *EventName;
+    uint64_t Instance;
+    size_t NumCategories;
+    std::array<const char *, 8> Categories;
+    trace_event_data_t *Parent;
+    trace_event_data_t *Event;
+    uint64_t Value;
+    size_t PID;
+    size_t TID;
+    size_t TimeStamp;
+  };
+
+  class TracePointBuffer {
+  public:
+    constexpr static size_t BUFFER_SIZE = 256;
+    TracePointBuffer() { MTracePoints.reserve(BUFFER_SIZE); }
+
+    bool isFull() const noexcept { return MTracePoints.size() == BUFFER_SIZE; }
+
+    void add(TracePoint &&tp) noexcept {
+      MTracePoints.push_back(std::forward<TracePoint>(tp));
+    }
+
+    std::vector<TracePoint>::const_iterator begin() const noexcept {
+      return MTracePoints.begin();
+    }
+
+    std::vector<TracePoint>::const_iterator end() const noexcept {
+      return MTracePoints.end();
+    }
+
+  private:
+    std::vector<TracePoint> MTracePoints;
+  };
+
+  TracePointBuffer *getOrCreateBuffer() {
+    SharedLock<SharedSpinLock> Lock{MBuffersMutex};
+    auto ID = std::this_thread::get_id();
+
+    if (MBuffers.count(ID) != 0)
+      return &MBuffers.at(ID);
+
+    Lock.upgrade_to_writer();
+
+    MBuffers.insert({ID, TracePointBuffer()});
+
+    return &MBuffers.at(ID);
```

Comment:
Style comment: a bit shorter and less repetitive version:
```suggestion
    if (MBuffers.count(ID) == 0) {
      Lock.upgrade_to_writer();
      MBuffers.insert({ID, TracePointBuffer()});
    }

    return &MBuffers.at(ID);
```